File: src/resilience/retry.rs
//! Retry pattern implementation for Hivemind
//!
//! This module provides an implementation of the Retry pattern with exponential backoff,
//! which automatically retries failed operations with increasing delays between attempts.

use anyhow::Result;
use serde::{Deserialize, Serialize};
use std::sync::Arc;
use std::time::{Duration, SystemTime, UNIX_EPOCH};
use tokio::sync::Mutex;
use tracing::{debug, error, info, warn};
use rand::Rng;

/// Exponential backoff configuration
#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct ExponentialBackoffConfig {
    /// Initial delay in milliseconds
    pub initial_delay_ms: u64,
    /// Maximum delay in milliseconds
    pub max_delay_ms: u64,
    /// Maximum number of retries
    pub max_retries: u32,
    /// Multiplier for each retry
    pub multiplier: f64,
    /// Whether to add jitter to the delay
    pub jitter: bool,
    /// Maximum total duration for all retries in milliseconds
    pub max_duration_ms: Option<u64>,
}

impl Default for ExponentialBackoffConfig {
    fn default() -> Self {
        Self {
            initial_delay_ms: 100,
            max_delay_ms: 10000,
            max_retries: 3,
            multiplier: 2.0,
            jitter: true,
            max_duration_ms: Some(30000),
        }
    }
}

/// Retry policy metrics
#[derive(Clone, Debug, Serialize, Deserialize, Default)]
pub struct RetryMetrics {
    /// Total number of retry attempts
    pub total_retry_attempts: u64,
    /// Total number of successful retries
    pub successful_retries: u64,
    /// Total number of failed retries
    pub failed_retries: u64,
    /// Average delay between retries in milliseconds
    pub avg_delay_ms: f64,
    /// Maximum delay observed in milliseconds
    pub max_delay_observed_ms: u64,
    /// Average number of retries per operation
    pub avg_retries_per_operation: f64,
    /// Maximum number of retries observed for a single operation
    pub max_retries_observed: u32,
}

/// Retry policy implementation
#[derive(Clone)]
pub struct RetryPolicy {
    /// Service name
    service_name: String,
    /// Retry configuration
    config: Arc<Mutex<ExponentialBackoffConfig>>,
    /// Retry metrics
    metrics: Arc<Mutex<RetryMetrics>>,
}

/// Trait for errors that can be retried
pub trait RetryableError {
    /// Check if the error is retryable
    fn is_retryable(&self) -> bool {
        true
    }
    
    /// Create an error from a string
    fn from_error(msg: &str) -> Self;
}

impl RetryPolicy {
    /// Create a new retry policy
    pub fn new(service_name: String, config: ExponentialBackoffConfig) -> Self {
        Self {
            service_name,
            config: Arc::new(Mutex::new(config)),
            metrics: Arc::new(Mutex::new(RetryMetrics::default())),
        }
    }
    
    /// Get current metrics
    pub async fn get_metrics(&self) -> RetryMetrics {
        self.metrics.lock().await.clone()
    }
    
    /// Update configuration
    pub async fn update_config(&self, config: ExponentialBackoffConfig) {
        let mut current_config = self.config.lock().await;
        *current_config = config;
        
        info!(
            "Retry policy for service {} updated configuration (initial_delay_ms: {}, max_delay_ms: {}, max_retries: {})",
            self.service_name, 
            current_config.initial_delay_ms,
            current_config.max_delay_ms,
            current_config.max_retries
        );
    }
    
    /// Reset metrics
    pub async fn reset_metrics(&self) {
        *self.metrics.lock().await = RetryMetrics::default();
        
        info!("Retry policy for service {} reset metrics", self.service_name);
    }
    
    /// Calculate delay for a retry attempt
    async fn calculate_delay(&self, attempt: u32) -> Duration {
        let config = self.config.lock().await.clone();
        
        // Calculate exponential delay
        let base_delay_ms = (config.initial_delay_ms as f64 * config.multiplier.powi(attempt as i32)) as u64;
        
        // Cap at max delay
        let capped_delay_ms = base_delay_ms.min(config.max_delay_ms);
        
        // Add jitter if enabled
        let final_delay_ms = if config.jitter {
            let jitter_factor = rand::thread_rng().gen_range(0.8..1.2);
            (capped_delay_ms as f64 * jitter_factor) as u64
        } else {
            capped_delay_ms
        };
        
        // Update metrics
        let mut metrics = self.metrics.lock().await;
        
        // Update max delay observed
        if final_delay_ms > metrics.max_delay_observed_ms {
            metrics.max_delay_observed_ms = final_delay_ms;
        }
        
        // Update average delay
        let total_attempts = metrics.total_retry_attempts;
        if total_attempts > 0 {
            metrics.avg_delay_ms = ((metrics.avg_delay_ms * (total_attempts as f64)) + (final_delay_ms as f64)) / ((total_attempts + 1) as f64);
        } else {
            metrics.avg_delay_ms = final_delay_ms as f64;
        }
        
        Duration::from_millis(final_delay_ms)
    }
    
    /// Execute a function with retry
    pub async fn execute<F, Fut, T, E>(&self, f: F) -> Result<T, E>
    where
        F: Fn() -> Fut,
        Fut: std::future::Future<Output = Result<T, E>>,
        E: std::error::Error + RetryableError,
    {
        let config = self.config.lock().await.clone();
        let start_time = SystemTime::now();
        let mut attempt = 0;
        let mut last_error = None;
        
        loop {
            // Check if we've exceeded max retries
            if attempt > config.max_retries {
                // Update metrics
                let mut metrics = self.metrics.lock().await;
                metrics.failed_retries += 1;
                
                warn!(
                    "Retry policy for service {} exceeded maximum retries ({}) for operation",
                    self.service_name, config.max_retries
                );
                
                return Err(last_error.unwrap());
            }
            
            // Check if we've exceeded max duration
            if let Some(max_duration_ms) = config.max_duration_ms {
                let elapsed = SystemTime::now()
                    .duration_since(start_time)
                    .unwrap_or_default()
                    .as_millis() as u64;
                
                if elapsed > max_duration_ms {
                    // Update metrics
                    let mut metrics = self.metrics.lock().await;
                    metrics.failed_retries += 1;
                    
                    warn!(
                        "Retry policy for service {} exceeded maximum duration ({}ms) for operation",
                        self.service_name, max_duration_ms
                    );
                    
                    return Err(E::from_error(&format!("Exceeded maximum retry duration of {}ms", max_duration_ms)));
                }
            }
            
            // Execute the function
            let result = f().await;
            
            match result {
                Ok(value) => {
                    // Success, update metrics and return
                    let mut metrics = self.metrics.lock().await;
                    
                    if attempt > 0 {
                        metrics.successful_retries += 1;
                        
                        // Update average retries per operation
                        let total_operations = metrics.successful_retries + metrics.failed_retries;
                        if total_operations > 0 {
                            metrics.avg_retries_per_operation = (metrics.total_retry_attempts as f64) / (total_operations as f64);
                        }
                        
                        debug!(
                            "Retry policy for service {} succeeded after {} attempts",
                            self.service_name, attempt + 1
                        );
                    }
                    
                    return Ok(value);
                }
                Err(err) => {
                    // Check if the error is retryable
                    if !err.is_retryable() {
                        // Non-retryable error, update metrics and return
                        let mut metrics = self.metrics.lock().await;
                        metrics.failed_retries += 1;
                        
                        warn!(
                            "Retry policy for service {} encountered non-retryable error: {}",
                            self.service_name, err
                        );
                        
                        return Err(err);
                    }
                    
                    // Update metrics
                    let mut metrics = self.metrics.lock().await;
                    metrics.total_retry_attempts += 1;
                    
                    // Update max retries observed
                    if attempt + 1 > metrics.max_retries_observed {
                        metrics.max_retries_observed = attempt + 1;
                    }
                    
                    // Store the error for potential return
                    last_error = Some(err);
                    
                    // Increment attempt counter
                    attempt += 1;
                    
                    // Calculate delay for next retry
                    let delay = self.calculate_delay(attempt).await;
                    
                    debug!(
                        "Retry policy for service {} failed attempt {}/{}, retrying in {}ms",
                        self.service_name, attempt, config.max_retries, delay.as_millis()
                    );
                    
                    // Wait before retrying
                    tokio::time::sleep(delay).await;
                }
            }
        }
    }
}

/// Implementation of RetryableError for std::io::Error
impl RetryableError for std::io::Error {
    fn is_retryable(&self) -> bool {
        use std::io::ErrorKind;
        
        match self.kind() {
            // Network-related errors are typically retryable
            ErrorKind::ConnectionRefused |
            ErrorKind::ConnectionReset |
            ErrorKind::ConnectionAborted |
            ErrorKind::NotConnected |
            ErrorKind::AddrInUse |
            ErrorKind::AddrNotAvailable |
            ErrorKind::TimedOut |
            ErrorKind::Interrupted |
            ErrorKind::WouldBlock => true,
            
            // Other errors are generally not retryable
            _ => false,
        }
    }
    
    fn from_error(msg: &str) -> Self {
        std::io::Error::new(std::io::ErrorKind::Other, msg)
    }
}

/// Implementation of RetryableError for anyhow::Error
impl RetryableError for anyhow::Error {
    fn from_error(msg: &str) -> Self {
        anyhow::anyhow!("{}", msg)
    }
}

/// Implementation of RetryableError for Box<dyn std::error::Error>
impl RetryableError for Box<dyn std::error::Error + Send + Sync> {
    fn from_error(msg: &str) -> Self {
        Box::new(std::io::Error::new(std::io::ErrorKind::Other, msg))
    }
}

File: internal/resilience/retry.go
// Package resilience provides resilience patterns for the Hivemind system.
package resilience

import (
	"context"
	"math"
	"math/rand"
	"sync"
	"time"

	"go.uber.org/zap"
)

// ExponentialBackoffConfig defines the configuration for exponential backoff
type ExponentialBackoffConfig struct {
	// InitialDelayMs is the initial delay in milliseconds
	InitialDelayMs int64
	// MaxDelayMs is the maximum delay in milliseconds
	MaxDelayMs int64
	// MaxRetries is the maximum number of retries
	MaxRetries uint32
	// Multiplier is the multiplier for each retry
	Multiplier float64
	// Jitter indicates whether to add jitter to the delay
	Jitter bool
	// MaxDurationMs is the maximum total duration for all retries in milliseconds
	MaxDurationMs *int64
}

// DefaultExponentialBackoffConfig returns the default exponential backoff configuration
func DefaultExponentialBackoffConfig() *ExponentialBackoffConfig {
	maxDuration := int64(30000)
	return &ExponentialBackoffConfig{
		InitialDelayMs: 100,
		MaxDelayMs:     10000,
		MaxRetries:     3,
		Multiplier:     2.0,
		Jitter:         true,
		MaxDurationMs:  &maxDuration,
	}
}

// RetryMetrics contains metrics for a retry policy
type RetryMetrics struct {
	// TotalRetryAttempts is the total number of retry attempts
	TotalRetryAttempts uint64
	// SuccessfulRetries is the total number of successful retries
	SuccessfulRetries uint64
	// FailedRetries is the total number of failed retries
	FailedRetries uint64
	// AvgDelayMs is the average delay between retries in milliseconds
	AvgDelayMs float64
	// MaxDelayObservedMs is the maximum delay observed in milliseconds
	MaxDelayObservedMs int64
	// AvgRetriesPerOperation is the average number of retries per operation
	AvgRetriesPerOperation float64
	// MaxRetriesObserved is the maximum number of retries observed for a single operation
	MaxRetriesObserved uint32
}

// RetryPolicy implements the retry pattern with exponential backoff
type RetryPolicy struct {
	// ServiceName is the name of the service
	serviceName string
	// Config is the retry configuration
	config *ExponentialBackoffConfig
	// Metrics is the retry metrics
	metrics *RetryMetrics
	// Mutex for protecting metrics updates
	mu sync.Mutex
	// Logger
	logger *zap.Logger
	// Random number generator for jitter
	rand *rand.Rand
}

// NewRetryPolicy creates a new retry policy
func NewRetryPolicy(serviceName string, config ExponentialBackoffConfig) *RetryPolicy {
	logger, _ := zap.NewProduction()

	return &RetryPolicy{
		serviceName: serviceName,
		config:      &config,
		metrics:     &RetryMetrics{},
		logger:      logger,
		rand:        rand.New(rand.NewSource(time.Now().UnixNano())),
	}
}

// GetMetrics returns the current metrics
func (p *RetryPolicy) GetMetrics() RetryMetrics {
	p.mu.Lock()
	defer p.mu.Unlock()

	return *p.metrics
}

// UpdateConfig updates the retry configuration
func (p *RetryPolicy) UpdateConfig(config ExponentialBackoffConfig) {
	p.mu.Lock()
	defer p.mu.Unlock()

	p.config = &config

	p.logger.Info("Retry policy updated configuration",
		zap.String("service", p.serviceName),
		zap.Int64("initial_delay_ms", config.InitialDelayMs),
		zap.Int64("max_delay_ms", config.MaxDelayMs),
		zap.Uint32("max_retries", config.MaxRetries))
}

// ResetMetrics resets the metrics
func (p *RetryPolicy) ResetMetrics() {
	p.mu.Lock()
	defer p.mu.Unlock()

	p.metrics = &RetryMetrics{}

	p.logger.Info("Retry policy reset metrics", zap.String("service", p.serviceName))
}

// calculateDelay calculates the delay for a retry attempt
func (p *RetryPolicy) calculateDelay(attempt uint32) time.Duration {
	p.mu.Lock()
	defer p.mu.Unlock()

	// Calculate exponential delay
	baseDelayMs := float64(p.config.InitialDelayMs) * math.Pow(p.config.Multiplier, float64(attempt))

	// Cap at max delay
	cappedDelayMs := math.Min(baseDelayMs, float64(p.config.MaxDelayMs))

	// Add jitter if enabled
	finalDelayMs := cappedDelayMs
	if p.config.Jitter {
		jitterFactor := 0.8 + p.rand.Float64()*0.4 // Random value between 0.8 and 1.2
		finalDelayMs = cappedDelayMs * jitterFactor
	}

	// Update metrics
	if int64(finalDelayMs) > p.metrics.MaxDelayObservedMs {
		p.metrics.MaxDelayObservedMs = int64(finalDelayMs)
	}

	// Update average delay
	totalAttempts := p.metrics.TotalRetryAttempts
	if totalAttempts > 0 {
		p.metrics.AvgDelayMs = ((p.metrics.AvgDelayMs * float64(totalAttempts)) + finalDelayMs) / float64(totalAttempts+1)
	} else {
		p.metrics.AvgDelayMs = finalDelayMs
	}

	return time.Duration(finalDelayMs) * time.Millisecond
}

// Execute executes a function with retry
func (p *RetryPolicy) Execute(ctx context.Context, f func(ctx context.Context) (interface{}, error)) (interface{}, error) {
	var zero interface{}

	config := *p.config
	startTime := time.Now()
	var attempt uint32
	var lastErr error

	for {
		// Check if we've exceeded max retries
		if attempt > config.MaxRetries {
			p.mu.Lock()
			p.metrics.FailedRetries++
			p.mu.Unlock()

			p.logger.Warn("Retry policy exceeded maximum retries",
				zap.String("service", p.serviceName),
				zap.Uint32("max_retries", config.MaxRetries))

			if lastErr != nil {
				return zero, lastErr
			}
			return zero, ErrMaxRetriesReached
		}

		// Check if we've exceeded max duration
		if config.MaxDurationMs != nil {
			elapsed := time.Since(startTime).Milliseconds()
			if elapsed > *config.MaxDurationMs {
				p.mu.Lock()
				p.metrics.FailedRetries++
				p.mu.Unlock()

				p.logger.Warn("Retry policy exceeded maximum duration",
					zap.String("service", p.serviceName),
					zap.Int64("max_duration_ms", *config.MaxDurationMs))

				return zero, ErrMaxDurationReached
			}
		}

		// Execute the function
		result, err := f(ctx)

		if err == nil {
			// Success, update metrics and return
			p.mu.Lock()

			if attempt > 0 {
				p.metrics.SuccessfulRetries++

				// Update average retries per operation
				totalOperations := p.metrics.SuccessfulRetries + p.metrics.FailedRetries
				if totalOperations > 0 {
					p.metrics.AvgRetriesPerOperation = float64(p.metrics.TotalRetryAttempts) / float64(totalOperations)
				}

				p.logger.Debug("Retry policy succeeded",
					zap.String("service", p.serviceName),
					zap.Uint32("attempts", attempt+1))
			}

			p.mu.Unlock()

			return result, nil
		}

		// Check if the error is retryable
		var retryable bool
		if re, ok := err.(RetryableError); ok {
			retryable = re.IsRetryable()
		} else {
			// Default to retryable
			retryable = true
		}

		if !retryable {
			// Non-retryable error, update metrics and return
			p.mu.Lock()
			p.metrics.FailedRetries++
			p.mu.Unlock()

			p.logger.Warn("Retry policy encountered non-retryable error",
				zap.String("service", p.serviceName),
				zap.Error(err))

			return zero, err
		}

		// Update metrics
		p.mu.Lock()
		p.metrics.TotalRetryAttempts++

		// Update max retries observed
		if attempt+1 > p.metrics.MaxRetriesObserved {
			p.metrics.MaxRetriesObserved = attempt + 1
		}
		p.mu.Unlock()

		// Store the error for potential return
		lastErr = err

		// Increment attempt counter
		attempt++

		// Calculate delay for next retry
		delay := p.calculateDelay(attempt)

		p.logger.Debug("Retry policy failed attempt, retrying",
			zap.String("service", p.serviceName),
			zap.Uint32("attempt", attempt),
			zap.Uint32("max_retries", config.MaxRetries),
			zap.Int64("delay_ms", delay.Milliseconds()))

		// Create a timer for the delay
		timer := time.NewTimer(delay)

		// Wait for either the delay to expire or the context to be canceled
		select {
		case <-timer.C:
			// Continue with the next retry
		case <-ctx.Done():
			// Context was canceled
			timer.Stop()

			p.mu.Lock()
			p.metrics.FailedRetries++
			p.mu.Unlock()

			p.logger.Warn("Retry policy context canceled",
				zap.String("service", p.serviceName),
				zap.Error(ctx.Err()))

			return zero, ctx.Err()
		}
	}
}

// StandardRetryableError is a standard implementation of RetryableError
type StandardRetryableError struct {
	error
	retryable bool
}

// IsRetryable returns whether the error is retryable
func (e StandardRetryableError) IsRetryable() bool {
	return e.retryable
}

// NewRetryableError creates a new retryable error
func NewRetryableError(err error, retryable bool) RetryableError {
	return StandardRetryableError{
		error:     err,
		retryable: retryable,
	}
}

// IsNetworkError checks if an error is a network-related error
func IsNetworkError(err error) bool {
	// In a real implementation, this would check for specific network error types
	// For now, we'll just return true for demonstration purposes
	return true
}

// IsTimeoutError checks if an error is a timeout error
func IsTimeoutError(err error) bool {
	// In a real implementation, this would check for specific timeout error types
	// For now, we'll just return true for demonstration purposes
	return true
}

// IsRetryableError checks if an error is retryable
func IsRetryableError(err error) bool {
	// Check if the error implements RetryableError
	if re, ok := err.(RetryableError); ok {
		return re.IsRetryable()
	}

	// Check for common retryable errors
	return IsNetworkError(err) || IsTimeoutError(err)
}
